Src/Newtonsoft.Json/JsonSerializer.cs

```diff
@@ -1187,28 +1187,32 @@ internal IReferenceResolver GetReferenceResolver()
             return _referenceResolver;
         }
 
-        internal JsonConverter GetMatchingConverter(Type type)
+        internal JsonConverter GetMatchingConverter(Type type, bool forWriting)
         {
-            return GetMatchingConverter(_converters, type);
+            return GetMatchingConverter(_converters, type, forWriting);
         }
 
-        internal static JsonConverter GetMatchingConverter(IList<JsonConverter> converters, Type objectType)
+        internal static JsonConverter GetMatchingConverter(IList<JsonConverter> converters, Type objectType, bool forWriting)
         {
 #if DEBUG
             ValidationUtils.ArgumentNotNull(objectType, nameof(objectType));
 #endif
 
             if (converters != null)
             {
+                JsonConverter possibleConverter = null;
                 for (int i = 0; i < converters.Count; i++)
                 {
                     JsonConverter converter = converters[i];
 
                     if (converter.CanConvert(objectType))
                     {
-                        return converter;
+                        if (forWriting ? converter.CanWrite : converter.CanRead)
+                            return converter;
+                        possibleConverter = converter;
```

Comment:
If there is no suitable converter for writing or for reading, shouldnt we pick the first converter that can convert the type? This code looks like it picks the last one.

Reply:
There's a case for either, this was just less code :-)

Reply:
Allright, but the old behavior is to choose the first one, I think. So it might be a good idea to keep that, to make the change footprint as small as possible and break fewer implementations.